pkg/gc/finalizer.go

```diff
@@ -149,29 +146,41 @@ func (f *subscriptionFinalizer) checkGeneration(ctx context.Context, typ core_mo
 			continue
 		}
 
-		old, ok := f.insights[tenantID(tenantId)][typ][key]
-		if !ok || old.id != insight.GetLastSubscription().GetId() || old.generation != insight.GetLastSubscription().GetGeneration() {
-			// something changed since the last check, either subscriptionId or generation were updated
-			// don't finalize the subscription, update map with fresh data
-			f.insights[tenantID(tenantId)][typ][key] = &descriptor{
-				id:         insight.GetLastSubscription().GetId(),
-				generation: insight.GetLastSubscription().GetGeneration(),
+		lastGens := f.insights[tenantID(tenantId)][typ][key]
+
+		subsToFinalize := map[string]struct{}{}
+		newWatchedSubs := subscriptionGenerationMap{}
+
+		for _, sub := range insight.AllSubscriptions() {
+			if !sub.IsOnline() {
+				continue
+			}
+			id := sub.GetId()
+			if gen, ok := lastGens[id]; !ok || gen != sub.GetGeneration() {
+				newWatchedSubs[id] = sub.GetGeneration()
+			} else {
+				subsToFinalize[id] = struct{}{}
 			}
-			continue
 		}
 
-		log.V(1).Info("mark subscription as disconnected")
-		insight.GetLastSubscription().SetDisconnectTime(now)
-
 		upsertInsight, _ := registry.Global().NewObject(typ)
-		err := manager.Upsert(ctx, f.rm, key, upsertInsight, func(r core_model.Resource) error {
-			return upsertInsight.GetSpec().(generic.Insight).UpdateSubscription(insight.GetLastSubscription())
-		})
-		if err != nil {
-			log.Error(err, "unable to finalize subscription")
+		if err := manager.Upsert(ctx, f.rm, key, upsertInsight, func(r core_model.Resource) error {
+			insight := upsertInsight.GetSpec().(generic.Insight)
+			for id := range subsToFinalize {
+				log.V(1).Info("mark subscription as disconnected", "id", id)
+				sub := insight.GetSubscription(id)
+				sub.SetDisconnectTime(now)
+				if err := insight.UpdateSubscription(sub); err != nil {
+					return errors.Wrapf(err, "unable to update subscription %s", id)
+				}
+			}
+			return nil
+		}); err != nil {
+			log.Error(err, "unable to upsert insight")
```

Comment:
I know this was already like this but

This should be wrap not log + err. Also the log has no tenant context.

Reply:
you meant return the wrapped error and don't log right? fixed